### PR TITLE
Support for Wasm ABI

### PR DESCRIPTION
**Pull request by bradlhart**
## Change Description
Wasm ABIs are a new format of abis that expose an interface that allows actions to be called as functions.  This enables greater capacity to handle the data types in the contract, some of which couldn’t be serialized via ABIs.

Additionally, with the new Wasm ABI classes, a new short-hand design of constructing transactions was added to utilize the interfaces.  `TransactionBuilder`/`ActionBuilder`/`ActionSerializer` work together to allow users to create either a concise action/context-free action in `api.transact()` or a stateful transaction object that can be constructed and sent, using those concise actions.

Please note that these changes do not affect current implementations of `eosjs` and Json Abis can utilize the new systems.

## API Changes
- [x] API Changes

Additions:
`WasmAbi`: Class for holding the WASM module and creating the interface for actions
`WasmAbiProvider`: Class for storing and initializing WasmAbi class objects
`TransactionBuilder/ActionBuilder/ActionSerializer`:  Collection of classes to support a new design of creating transactions
`api.query()`: Function for new `queryit` contract to facilitate new wasmQL querying with rodeos 
New interfaces/functions for serializing `api.query()` requests and responses

Changes:
`api.transact()`: Optional argument `requiredKeys`

## Documentation Additions
- [x] Documentation Additions

Documentation detailing the new Wasm ABI objects, examples detailing the new short-hand design of transactions.

### CODE REVIEW

**Comment by GreenBusDriver** [comment]
Brad, why are there entire files that are new?  the token abi and wasm, the boot abi and wasm.  Are these new additions?

**Comment by bradlhart** [comment]
> Brad, why are there entire files that are new? the token abi and wasm, the boot abi and wasm. Are these new additions?

They are specifically for building the docker image for ci/cd

File: src/eosjs-api.ts [hunk 1]
```diff
@@ -317,14 +351,101 @@ export class Api {
             });
         }
         if (broadcast) {
-            if (compression) {
-                return this.pushCompressedSignedTransaction(pushTransactionArgs);
+            const result = await this.rpc.send_transaction(pushTransactionArgs);
```
Comment (tbfleming) [inline src/eosjs-api.ts:354]: What happened to `pushCompressedSignedTransaction`?

File: src/eosjs-api.ts [hunk 2]
```diff
@@ -216,16 +229,31 @@ export class Api {
...
     public async serializeActions(actions: ser.Action[]): Promise<ser.SerializedAction[]> {
-        return await Promise.all(actions.map(async ({ account, name, authorization, data }) => {
+        return await Promise.all(actions.map(async (action) => {
+            const { account, name, authorization, data } = action;
+            if (this.wasmAbiProvider && this.wasmAbiProvider.wasmAbis.get(account)) {
+                const wasmAbi = this.wasmAbiProvider.wasmAbis.get(account);
+                if (wasmAbi.inst.exports.memory.buffer.length > wasmAbi.memoryThreshold) {
+                    await wasmAbi.reset();
+                }
+                return action;
```
Comment (tbfleming) [inline src/eosjs-api.ts:239]: Where does it serialize?

File: src/eosjs-api.ts [hunk 3]
```diff
@@ -216,16 +229,31 @@ export class Api {
...
         }));
     }
 
     /** Convert actions from hex */
     public async deserializeActions(actions: ser.Action[]): Promise<ser.Action[]> {
-        return await Promise.all(actions.map(async ({ account, name, authorization, data }) => {
+        return await Promise.all(actions.map(async (action) => {
+            const { account, name, authorization, data } = action;
+            if (this.wasmAbiProvider && this.wasmAbiProvider.wasmAbis.get(account)) {
+                return action;
```
Comment (tbfleming) [inline src/eosjs-api.ts:255]: Where does it deserialize?

File: src/eosjs-api.ts (hunk 2 above)
Comment (tbfleming) [reply src/eosjs-api.ts:239]: Never mind. I see it now.

File: src/eosjs-api.ts (hunk 3 above)
Comment (tbfleming) [reply src/eosjs-api.ts:255]: Never mind. I see it now.

File: src/eosjs-serialize.ts [hunk 4]
```diff
@@ -1143,3 +1175,173 @@ export function deserializeAction(contract: Contract, account: string, name: str
...
+    int16: { index: 5, useShortForm: false, type: additionalTypes.get('int16') },
+    uint16: { index: 6, useShortForm: false, type: additionalTypes.get('uint16') },
+    int8: { index: 7, useShortForm: false, type: additionalTypes.get('int8') },
+    uint8: { index: 8, useShortForm: false, type: additionalTypes.get('uint8') },
+    time_point: { index: 9, useShortForm: false, type: additionalTypes.get('time_point') },
+    checksum256: { index: 10, useShortForm: false, type: additionalTypes.get('checksum256') },
+    float64: { index: 11, useShortForm: false, type: additionalTypes.get('float64') },
+    string: { index: 12, useShortForm: true, type: additionalTypes.get('string') },
+    any_object: { index: 13, useShortForm: true, type: additionalTypes.get('any_object') },
+    any_array: { index: 14, useShortForm: true, type: additionalTypes.get('any_array') },
```
Comment (tbfleming) [inline src/eosjs-serialize.ts:1216]: More types were added since this list was generated

Comment (bradlhart) [reply src/eosjs-serialize.ts:1226]: Added the new types

File: src/eosjs-api.ts (hunk 1 above)
Comment (bradlhart) [reply src/eosjs-api.ts]: Reverted that change

File: src/eosjs-serialize.ts [hunk 5]
```diff
@@ -1193,6 +1193,16 @@ function addAdditionalTypes(): Map<string, Type> {
         serialize: serializeAnyArray,
         deserialize: deserializeAnyArray
     }));
+    initialTypes.set('any_object', createType({
```
Comment (tbfleming) [inline src/eosjs-serialize.ts:1196]: duplicate

File: src/eosjs-serialize.ts [hunk 6]
```diff
@@ -1193,6 +1193,16 @@ function addAdditionalTypes(): Map<string, Type> {
...
+        name: 'any_object',
+        serialize: serializeAnyObject,
+        deserialize: deserializeAnyObject
+    }));
+    initialTypes.set('any_array', createType({
```
Comment (tbfleming) [inline src/eosjs-serialize.ts:1201]: duplicate

Comment (bradlhart) [reply src/eosjs-serialize.ts]: removed

File: src/eosjs-serialize.ts (hunk 5 above)
Comment (bradlhart) [reply src/eosjs-serialize.ts]: removed

**Pull request merged**